notifications/manager.go

```diff
@@ -0,0 +1,205 @@
+package notifications
+
+import (
+	"context"
+	"time"
+
+	"github.com/lightninglabs/loop/swapserverrpc"
+	"google.golang.org/grpc"
+)
+
+type NotificationType int
+
+const (
+	NotificationTypeUnknown NotificationType = iota
+	NotificationTypeReservation
+)
+
+type NotificationsClient interface {
+	SubscribeNotifications(ctx context.Context,
+		in *swapserverrpc.SubscribeNotificationsRequest,
+		opts ...grpc.CallOption) (
+		swapserverrpc.SwapServer_SubscribeNotificationsClient, error)
+}
+
+// Config contains all the services that the notification manager needs to
+// operate.
+type Config struct {
+	// Client is the client used to communicate with the swap server.
+	Client NotificationsClient
+
+	// FetchL402 is the function used to fetch the l402 token.
+	FetchL402 func(context.Context) error
+}
+
+// Manager is a manager for notifications that the swap server sends to the
+// client.
+type Manager struct {
+	cfg *Config
+
+	hasL402 bool
+
+	subscribers map[NotificationType][]subscriber
+}
+
+// NewManager creates a new notification manager.
+func NewManager(cfg *Config) *Manager {
+	return &Manager{
+		cfg:         cfg,
+		subscribers: make(map[NotificationType][]subscriber),
+	}
+}
+
+type subscriber struct {
+	subCtx   context.Context
+	recvChan interface{}
+}
+
+// SubscribeReservations subscribes to the reservation notifications.
+func (m *Manager) SubscribeReservations(ctx context.Context,
+) <-chan *swapserverrpc.ServerReservationNotification {
+
+	// We'll create a channel that we'll use to send the notifications to the
+	// caller.
+	notifChan := make(
+		chan *swapserverrpc.ServerReservationNotification, //nolint:lll
+	)
+	sub := subscriber{
+		subCtx:   ctx,
+		recvChan: notifChan,
+	}
+
+	m.subscribers[NotificationTypeReservation] = append(
+		m.subscribers[NotificationTypeReservation],
+		sub,
+	)
+
+	return notifChan
+}
+
+// Run starts the notification manager.
+func (n *Manager) Run(ctx context.Context) error {
+	err := n.registerNotifications(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// registerNotifications registers a new server notification stream.
+func (n *Manager) registerNotifications(mainCtx context.Context) error {
+	// In order to create a valid l402 we first are going to call
+	// the FetchL402 method. As a client might not have outbound capacity
+	// yet, we'll retry until we get a valid response.
+	if !n.hasL402 {
+		n.fetchL402(mainCtx)
+	}
+
+	callCtx, cancel := context.WithCancel(mainCtx)
+
+	// We'll now subscribe to the swap server notifications.
+	notifStream, err := n.cfg.Client.SubscribeNotifications(
+		callCtx, &swapserverrpc.SubscribeNotificationsRequest{},
+	)
+	if err != nil {
+		cancel()
```

Comment:
Add an empty line

Reply:
I agree, it makes sense not to add a new line here.

On the other hand, manually called `cancel` (not in `defer`) looks suspicious. I think, it won't be needed if the code is reorganized into a single event loop (and goroutine).